datalad_ria/patches/sshremoteio.py

```diff
@@ -0,0 +1,149 @@
+"""
+
+This patch fixes errors in the __init__ function of
+datalad.distributed.ora_remote.py:SSHRemoteIO.
+
+The original code has two problems.
+
+1. The `cmd`-argument for the shell ssh-process, which is created by:
+   `self.shell = subprocess.Popen(cmd, ...)`
+   is not correct, if `self.ssh` is an instance of `NoMultiplexSSHConnection`.
+
+   The changes in this patch build the correct `cmd`-argument by adding
+   additional parameters to `cmd`, if `self.ssh` is an instance of
+   `NoMultiplexSSHConnection`.
```

Comment:
Thanks much for the description!

Looking at the code and this description, I wonder if we can add some idea on a direction to rectify this *within the scope of the RIA functionality*.

I assume that "build the correct `cmd`-argument by adding additional parameters to `cmd`" means that some information is dropped somewhere (would be good to make that explicit and mention what it is).

From the RIA perspective all information is in the URL that we pass into the `SSHConnector` code stack. Internally, it is parsed to an `SSHRI`, which is then assembled into the `ssh` command call, which uses a different style of argument passing (i.e., SSH can handle `ssh://` URL perfectly fine, but datalad goes the disassemble/reassemble route nevertheless, and fails trying).

In order to make this patch obsolete eventually, there appear to be two different approaches:

1. Adopt this patch upstream. IOW fix the disassemble/reassemble approach
2. Adjust `SSHConnector` to be able to work with SSH-URLs directly/verbatim

If that is correct and a comprehensive list, which direction would be "better"?

Reply:
> I assume that "build the correct cmd-argument by adding additional parameters to cmd" means that some information is dropped somewhere (would be good to make that explicit and mention what it is).

The complete description is complex and involves knowledge of the internals of the ``SSHManager``-classes, more precisely,  knowledge of the different semantics of "opening a connection" in ``NoMultiplexSSHManager`` and ``MultiplexSSHConnection``. That leads to the situation that the necessary ``ssh``-paramater to open a shell are either stored in ``BaseSSHConnection._ssh_open_args`` or in ``BaseSSHConnection._ssh_args``. This is my understanding of the situation (the original authors of the code might understand this differently). @yarikoptic verified the approach of this patch. I added a "short"-version of this explanation in d53999807363a8012fd831436fca2f1f7077699f.


> From the RIA perspective all information is in the URL that we pass into the SSHConnector code stack. Internally, it is parsed to an SSHRI, which is then assembled into the ssh command call, which uses a different style of argument passing (i.e., SSH can handle ssh:// URL perfectly fine, but datalad goes the disassemble/reassemble route nevertheless, and fails trying).

SSHRI does not really fail. There are parameters that cannot be represented in an SSHRI nor in a URL. The problems that this PR solves are due to the fact that the parameters, which are not represented in the SSHRI, were not provided as arguments to an ``ssh``-subprocess call. More precisely, the port and the identity-file were not provided. While sticking to URLs would provide the port, the identity file would still have to be given to the ``ssh``-subprocess as an additional argument. Having said that, I think we should try to switch to URLs and existing libraries.

> In order to make this patch obsolete eventually, there appear to be two different approaches:
> 1. Adopt this patch upstream. IOW fix the disassemble/reassemble approach
> 2. Adjust SSHConnector to be able to work with SSH-URLs directly/verbatim
> If that is correct and a comprehensive list, which direction would be "better"?

As outlined above, the problems solved by this PR stem from faulty command line argument handling. So they are somewhat orthogonal to the URL vs. SSHRI discussion.

Answer 1. the patch has to go into the upstream code, because it will be required for SSHRI-based and URL-based SSHConnection code. The patch is independent of the assemble/reassemble-code and it does actually not touch this part of the code anymore.

Answer 2. as mentioned above, I think it would be better if we could get rid of the home-grown `RI` logic (at least for ssh). I find it to be complicated and it is a substantial amount of code that has to be maintained. But, it is not necessary to fix the issue of this PR, if the patch is applied upstream

Reply:
Ok, thanks. I was hoping to have this reflected in the actual code file rather than a github comment. but let's hope we never need to find it again (it will be in the wrong project once that patch is moved).

But let's move on.